### PR TITLE
supporting x dtype instead of only float 32 in K.relu

### PR DESCRIPTION
**Pull request by vishalsubbiah**
K.relu casts to floatx() by default when threshold is greater than 0. This fails when one sets mixed precision through `tf.keras.mixed_precision.experimental.set_policy` to `mixed_float16` or just pass an input of dtype float16

Changing it to follow the documentation to use the inputs dtype instead.

#40687
 cc @Saduf2019

### CODE REVIEW

**Review by qlzh727 — DISMISSED** [review]

**Comment by vishalsubbiah** [comment]
@qlzh727  fixed the tests. Can you review again please?

**Review by qlzh727 — DISMISSED** [review]

File: tensorflow/python/keras/backend.py [hunk 1]
```diff
@@ -4541,7 +4541,13 @@ def relu(x, alpha=0., max_value=None, threshold=0):
   Returns:
       A tensor.
   """
-
+  if isinstance(x, (ops.Tensor,
+                    variables_module.Variable,
+                    sparse_tensor.SparseTensor,
+                    np.ndarray)):
+    dtype = x.dtype
```
Comment (qlzh727) [inline tensorflow/python/keras/backend.py:4548]: Just curious why we need to do the type check here, shouldn't it always use the input dtype?

Comment (vishalsubbiah) [reply tensorflow/python/keras/backend.py:4548]: I thought so too, but in some of the tests for this, they pass in a list/tuple which doesn't have a dtype defined, so in those cases we default to float32.

Comment (qlzh727) [reply tensorflow/python/keras/backend.py:4548]: I see. Then should we just check if 'x' has dtype attribute? I would expect the list of types here to grow when we have more types. Have to main the list here will be a maintenance headache.

Comment (vishalsubbiah) [reply tensorflow/python/keras/backend.py:4545]: @qlzh727 changed it to check if the dtype attribute is available instead.

Comment (qlzh727) [reply tensorflow/python/keras/backend.py:4545]: Thanks. How about:

```
dtype = getattr(x, 'dtype', floatx())
```

Comment (qlzh727) [reply tensorflow/python/keras/backend.py:4545]: Please also add a comment for those x input which doesn't have dtype, eg raw list and tuple.

Comment (vishalsubbiah) [reply tensorflow/python/keras/backend.py]: updated @qlzh727

**Review by qlzh727 — APPROVED** [review]
Thanks for the update.

**Pull request merged**